manager/src/main/java/org/openremote/manager/rules/JsonRulesBuilder.java

```diff
@@ -286,15 +286,85 @@ void update(Map<String, Long> nextRecurAssetIdMap) {
                 results.put(true, matched);
                 results.put(false, unmatched);
 
-                unfilteredAssetStates.stream().collect(Collectors.groupingBy(AttributeInfo::getId)).forEach((id, states) -> {
-                    Set<AttributeInfo> matches = assetPredicate.apply(states);
-                    if (matches != null) {
-                        matched.addAll(matches);
-                        unmatched.addAll(states.stream().filter(matches::contains).collect(Collectors.toSet()));
-                    } else {
-                        unmatched.addAll(states);
-                    }
-                });
+                unfilteredAssetStates.stream()
+                    .collect(Collectors.groupingBy(AttributeInfo::getId))
+                    .forEach((id, states) -> {
+
+                        // If the assetDurationPredicate is set then we use that to to match the asset states
+                        if (assetDurationPredicate != null) {
+                            // Pair of the attribute info that match and the index of the duration map
+                            Set<Pair<AttributeInfo, Integer>> matches = assetDurationPredicate.apply(states);
+                            
+                            if (matches != null) {
+                                long currentTime = timerService.getCurrentTimeMillis();
+                                
+                                matches.forEach(pair -> {
```

Comment:
Rename `pair` to something more meaningful e.g. `attributeInfoAndMatchedPredicateIndex`